### PR TITLE
feat: ✨ add conventional commits, and gitmoji cli flags

### PR DESCRIPTION
**Pull request by NorkzYT**
Built on-top of the current PR of https://github.com/Nutlope/aicommits/pull/160.

Commands to use after forking repo and switching to this PR:
```
pnpm i
pnpm build
node ./dist/cli.mjs
```

New flag information:

Flag for conventional commits:
Entire flag =`--conventional`
Alias = `-c`

Flag for gitmoji:
Entire flag = `--gitmoji`
Alias = `-m`

These are listed as Booleans in the code, although I could not figure out how to allow the use of `true` or `false.

To use conventional commits and gitmoji, run the following command:
```
node ./dist/cli.mjs -c -m
```

Example output 1:

Command:
`node ./dist/cli.mjs`
```
Add prompt for commit message formatting in generateCommitMessage function

This commit adds a prompt for commit message formatting in the generateCommitMessage function. The prompt is added after the base prompt and before the gitMojiExtraContext. This change helps ensure that commit messages conform to a consistent format. The text-davinci-003 model has a token limit of 4000, and this change does not affect this limit.
```

Example output 2:

Command:
`node ./dist/cli.mjs -c`
```
feat(openai): add prompt for generating commit messages

This commit adds a prompt for generating commit messages using OpenAI's GPT-3 API. The prompt includes a base prompt, a commit message format prompt, and an extra context prompt using gitMoji. The text-davinci-003 model has a token limit of 4000, and the prompt was updated to include an empty line for better readability.
```

Example output 3:

Command:
`node ./dist/cli.mjs -c -m`
```
🚀 feat: add base prompt and gitmoji context to commit message generation

This commit updates the `generateCommitMessage` function to include the `basePrompt` and `gitMojiExtraContext` in the `prompt` variable. The `prompt` variable 
is used to generate the commit message. This change allows for more information to be included in the commit message, making it more descriptive and helpful. 
```

Currently, the way that the prompt uses gitmoji's is insufficient, thus I will work on a way to retrieve the gitmoji's from the official api, and parse the prompt.

### CODE REVIEW

**Comment by NorkzYT** [comment]
Thank you for the explanation, I was wondering why you had `<contentional commit type>: <gitmoji> <message>`, and now I know. I will revert back. I did make a way to get the Gitmoji from the API, although I am messing with the prompt since it's not always accurate with the conventional commit scope, such as `feat:`.

I would say it would be best to launch as experimental and add documentation.

**Comment by thijslimmen** [comment]
I also wonder if you run the tests after all your changes. a lot of them are now failing. Also I created some tests that should correspond to a certain conventional commit type. I also saw that you removed the extra context for conventional commit types. That was there on purpose for GPT to be more accurate on selecting the types.

**Comment by NorkzYT** [comment]
Strange, I ran `pnpm test`, and all 5 checks have passed.

I needed to temporarily remove the conventional commit types to experiment with a different prompt since it was very long. I will add it back.

**Comment by thijslimmen** [comment]
There are 16 tests now, so maybe some of them didn't run on your machine :)

**Comment by NorkzYT** [comment]
Understood, I am going to switch to a Linux system.

**Comment by NorkzYT** [comment]
@thijslimmen 

I will keep working on this later, or tomorrow. I re-added the extra context for conventional commit types, the gitmoji api, and changed the prompt.

**Comment by thijslimmen** [comment]
That's cool, with the Gitmoji API! I would like to contribute more to this PR. What is a good way of working together in this PR? Fork your branch and push or clone your fork and commit on that? I also have some time tomorrow.

**Comment by thijslimmen** [comment]
I like that your latest commit messages are already generated with the new flags 💯

**Comment by NorkzYT** [comment]
I can add you as a contributor if you do not mind so you can work on that branch without making a new fork.

**Comment by NorkzYT** [comment]
Added.

**Comment by NorkzYT** [comment]
Note:

Tomorrow I will work on `git diff` parsing since some files may contain a lot of changed code which may result in the following caused by only 1 file:

```
{
  "error": {
    "message": "This model's maximum context length is 4097 tokens. However, your messages resulted in 11434 tokens. Please reduce the length of the messages.",
    "type": "invalid_request_error",
    "param": "messages",
    "code": "context_length_exceeded"
  }
}
```

Roadmap has something similar that needs to be done.
`Add support for diffs greater than 200 lines by grabbing the diff per file, optional flag`

I do not recall correctly, although I may have read an issue or PR that I cannot seem to find anymore about parsing the diff into smaller chunks or by having the most important parts be imported into the OpenAI prompt to receive an accurate git commit and not allow errors such as the aforementioned one to occur again.

**Comment by thijslimmen** [comment]
I see that the generation is a bit buggy (by running the unit tests). I will focus this evening on making that more robust with the unit tests and refining the prompt.

**Comment by NorkzYT** [comment]
Understood, thank you!

**Comment by NorkzYT** [comment]
The commit message description is excellent.

**Comment by thijslimmen** [comment]
I removed the token package, this can already be evaluated using the `openai` package. also doing the api call through the package instead of the custom post. Also improved on prompting and refactored separation for concern (openai) shouldnt be responsible for creating the git message string. 

Many improvements in general :)

**Comment by thijslimmen** [comment]
Larger staged files can now be provided, but still I am reaching a limit eventually

**Comment by thijslimmen** [comment]
I am really happy about the current PR. I will improve on unit testing this evening. Shall we then think about releasing it? 🚀

**Comment by NorkzYT** [comment]
> I am really happy about the current PR. I will improve on unit testing this evening (GMT+1). Shall we then think about releasing it? 🚀

I agree. Although before that I would like to work on adding the `GPT4` model option in this GitHub PR.

File: src/cli.ts [hunk 1]
```diff
@@ -24,6 +24,16 @@ cli(
 				description: 'Number of messages to generate. (Warning: generating multiple costs more) (default: 1)',
 				alias: 'g',
 			},
+			gitmoji: {
+				type: Boolean,
+				description: '(Experimental) Enable the use of Gitmoji',
+				alias: 'm',
+			},
```
[the remark this answers is not in the document]
Comment (NorkzYT) [reply src/cli.ts:31]: I had forgotten about that, how about we change the Gitmoji alias to `-gm`, and Conventional Commits to `-cc`?

**Comment by thijslimmen** [comment]
In my opinion we should complete this with GPT3.5. Also GPT4 is not publicly available and heavily throttled, so for now it is a nice to have which we can implement in a new PR

**Comment by NorkzYT** [comment]
Understood.

**Comment by thijslimmen** [comment]
@NorkzYT I've done the following things.

- Added `vitest` as unit test library which makes it possible to run tests concurrently, really fast feedback now
- Added unit tests for just the `openai.ts` class. I am using ChatGPT to generate a git diff based on some description to validate if correct conventional types are being used. This works really good! also pretty easy to setup a test case.

You should run, tests are really fast 💯 : 
```
pnpm build; OPENAI_KEY=<OPENAI_KEY>pnpm test
```

I do see some things that are not working too often:
- Gitmoji is provided 70% of the time
- When using conventional commits, the japanese language is sometimes not applied, also sometimes without conventional commits

I think the above happens, because the prompt is getting bigger and the subtle context is lost. We have to find a way to add some weight to certain things to make sure it is always applied.

We could also use a Semantic kernal pattern which validates the output and does some retrying until a certain condition is meet.

https://github.com/microsoft/semantic-kernel

**Comment by privatenumber** [comment]
Thanks for submitting the PR! I appreciate the effort you've put into making these changes. While some of them are quite beneficial, a few may require further discussion before we can proceed.

To smoothen the review process, would you mind dividing up the changes into smaller PRs? For each PR, please provide a brief explanation of the issue you're addressing, as this will offer context for review.

**Comment by thijslimmen** [comment]
@privatenumber Sure! I will create some smaller PRs. I think that will help you review the changes

**Comment by NorkzYT** [comment]
@thijslimmen 

Great work mate!

I will look into the prompt issue to find a way to weigh the prompt to increase accuracy of both the issues you mentioned.

**Comment by privatenumber** [comment]
Sounds good! Just so we're on the same page, smaller PRs ideally consist of 1~3 files. That should streamline review to deploy and help get your contributions in faster.

Additionally, it would be helpful if you could separate any large or non-essential changes, such as refactors, into a separate PR. This way, we can ensure that high-priority features are not delayed by lower-priority modifications.

Thanks again, your contribution is greatly appreciated!

**Comment by thijslimmen** [comment]
I continued in PR #168 

The Source code is now +- 4 files (including config files). Only code is updated that is related to supporting conventional commits.

I did improve a lot on the testing strategy, because it was really hard to test my changes. These changes are a bit bigger, but these should make reviewing easier.

I hope this helps :)

**Comment by thijslimmen** [comment]
I will close this PR. It will remain as reference for future changes like Gitmoji and Commit Title + Description

**Pull request closed without merging**